cache/cache_reservation_manager.h

```diff
@@ -172,20 +228,75 @@ class CacheReservationManager
   CacheKey cache_key_;
 };
 
-// CacheReservationHandle is for managing the lifetime of a cache reservation
-// This class is NOT thread-safe
-template <CacheEntryRole R>
-class CacheReservationHandle {
+// Same as CacheReservationManager, except that
+// it's thread-safe in calling all the APIs excluding construction and
+// destruction.
+//
+// See CacheReservationManager for more API usage.
+class CacheReservationManagerThreadSafeWrapper
+    : public std::enable_shared_from_this<
+          CacheReservationManagerThreadSafeWrapper> {
  public:
-  // REQUIRES: cache_res_mgr != nullptr
-  explicit CacheReservationHandle(
+  // Same as CacheReservationManager::CacheReservationHandle
+  // expect that it's for CacheReservationManagerThreadSafeWrapper
+  //
+  // This class is NOT thread-safe
+  //
+  // See CacheReservationManager::CacheReservationManager for more API usage.
+  template <CacheEntryRole R>
+  class CacheReservationHandle {
+   public:
+    // REQUIRES: cache_res_mgr_threadsafe_wrapper != nullptr
+    explicit CacheReservationHandle(
+        std::size_t incremental_memory_used,
+        std::shared_ptr<CacheReservationManagerThreadSafeWrapper>
+            cache_res_mgr_threadsafe_wrapper);
+
+    ~CacheReservationHandle();
+
+    std::size_t GetIncrementalMemoryUsed() { return incremental_memory_used_; }
+
+   private:
+    std::size_t incremental_memory_used_;
+    std::shared_ptr<CacheReservationManagerThreadSafeWrapper>
+        cache_res_mgr_threadsafe_wrapper_;
+  };
```

Comment:
Wait.. Now there are two `CacheReservationHandle` classes. How about making CacheReservationManagerThreadSafeWrapper a subclass of CacheReservationManager (alternative name idea: "ConcurrentCacheReservationManager") instead of a wrapper so we only need one?

Reply:
Interesting .. I was trying a similar thing to consolidate two CacheReseravationHandle but couldn't quite remember what didn't work. Let me try again harder

Reply:
As a subclass I'd just implement the methods like (i.e., not as a wrapper anymore):

```
... ConcurrentCacheReservationManager::ReleaseCacheReservation(...) {
  std::lock_guard<std::mutex> lock(mu_);
  return CacheReservationManager::ReleaseCacheReservation(...);
}
```

Reply:
IIUC, ReleaseCacheReservation is now protected so in a ConcurrentCacheReservationManager object, when trying to do something like cache_res_mgr_->ReleaseCacheReservation on the stored object cache_res_mgr_, it can not access the protected member of that object cache_res_mgr_. 

I have to duplicate the impl of CacheReservationManagerImpl::ReleaseCacheReservation ConcurrentCacheReservationManager to get around with this unfortunately. See note on the PR (attaching it ..)